File: solutions/go/triangle/1/triangle.go
// Package triangle should have a package comment that summarizes what it's about.
package triangle

import "math"

// Kind of triangle.
type Kind int

const (
	// NaT is not a triangle.
	NaT Kind = iota
	// Equ is equilateral.
	Equ
	// Iso is isosceles.
	Iso
	// Sca is scalene.
	Sca
)

// KindFromSides should have a comment documenting it.
func KindFromSides(a, b, c float64) Kind {
	if a <= 0 || b <= 0 || c <= 0 || a > b+c || b > a+c || c > a+b {
		return NaT
	}
	if math.IsNaN(a) || math.IsNaN(b) || math.IsNaN(c) || math.IsInf(a, 0) || math.IsInf(b, 0) || math.IsInf(c, 0) {
		return NaT
	}

	if a == b && b == c {
		return Equ
	} else if a == b || a == c || b == c {
		return Iso
	} else {
		return Sca
	}
}

File: solutions/cpp/triangle/1/triangle.cpp
#include "triangle.h"

#include <stdexcept>

namespace triangle {

flavor kind(double a, double b, double c)
{
    if (a <= 0 || b <= 0 || c <= 0) {
        throw std::domain_error("Triangle with zero or negative sides are illegal");
    }
    if (a + b < c || a + c < b || b + c < a) {
        throw std::domain_error("Triangle violating triangle inequality are illegal");
    }

    if (a == b && a == c) {
        return flavor::equilateral;
    }
    if (a == b || a == c || b == c) {
        return flavor::isosceles;
    }
    return flavor::scalene;
}

}  // namespace triangle
